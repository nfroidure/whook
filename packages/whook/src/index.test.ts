import { describe, test, jest, expect } from '@jest/globals';
import { constant } from 'knifecycle';
import { type Logger } from 'common-services';
import {
  runProcess,
  prepareProcess,
  prepareEnvironment as basePrepareEnvironment,
} from './index.js';

describe('runProcess', () => {
  test('should work', async () => {
    const PORT = 8888;
    const HOST = 'localhost';
    const BASE_PATH = '/v1';
    const API = {
      openapi: '3.1.0',
      info: {
        version: '1.0.0',
        title: 'Sample OpenAPI',
        description: 'A sample OpenAPI file for testing purpose.',
      },
      paths: {
        '/ping': {
          head: {
            operationId: 'getPing',
            summary: "Checks API's availability.",
            responses: {
              '200': {
                description: 'Pong',
              },
            },
          },
        },
      },
    };
    const logger = {
      output: jest.fn(),
      error: jest.fn(),
      debug: jest.fn(),
    };

    async function prepareEnvironment() {
      const $ = await basePrepareEnvironment();

      $.register(constant('BASE_PATH', BASE_PATH));
      $.register(constant('API', API));
      $.register(constant('APP_ENV', 'local'));
      $.register(
        constant('ENV', {
          NODE_ENV: 'test',
          JWT_SECRET: 'lol',
        }),
      );
      $.register(constant('PORT', PORT));
      $.register(constant('HOST', HOST));
      $.register(constant('WRAPPERS', []));
      $.register(constant('WHOOK_PLUGINS', []));
      $.register(constant('APP_CONFIG', {}));
      $.register(constant('WHOOK_RESOLVED_PLUGINS', []));
      $.register(constant('HTTP_SERVER_OPTIONS', {}));
      $.register(constant('DEBUG_NODE_ENVS', []));
      $.register(
        constant('HANDLERS', {
          getPing: jest.fn(() => ({ status: 200 })),
        }),
      );
      $.register(constant('logger', logger as Logger));
      $.register(constant('MAIN_FILE_URL', import.meta.url));

      return $;
    }
    process.env.ISOLATED_ENV = '1';

    const { $instance } = await runProcess(
      prepareEnvironment,
      prepareProcess,
      ['$instance', 'httpServer', 'process'],
      [],
    );

    await $instance.destroy();

    expect({
      debugCalls: logger.debug.mock.calls.sort(sortLogs),
      logInfoCalls: logger.output.mock.calls,
      logErrorCalls: logger.error.mock.calls,
    }).toMatchSnapshot();
  });
});

function sortLogs(strs1, strs2) {
  return strs1[0] > strs2[0] ? 1 : strs1[0] === strs2[0] ? 0 : -1;
}
