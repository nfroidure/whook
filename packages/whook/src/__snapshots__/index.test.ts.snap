// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runProcess should work 1`] = `
{
  "debugCalls": [
    [
      "⌛ - Delay service initialized.",
    ],
    [
      "⏰ - Time service initialized.",
    ],
    [
      "⏳ - Cancelling pending timeouts:",
      0,
    ],
    [
      "✅ - Closing HTTP server.",
    ],
    [
      "✔️ - HTTP server closed!",
    ],
    [
      "❤️ - Initializing the APM service.",
    ],
    [
      "👣 - Logging service initialized.",
    ],
    [
      "💱 - HTTP Transaction initialized.",
    ],
    [
      "📇 - Process service initialized.",
    ],
    [
      "🕶️ - Initializing the obfuscator service.",
    ],
    [
      "🚦 - HTTP Router initialized.",
    ],
  ],
  "logErrorCalls": [
    [
      "🎙️ - HTTP Server listening at "http://localhost:8888".",
    ],
    [
      "On air 🚀🌕",
    ],
  ],
  "logInfoCalls": [],
}
`;
